File: reactjs/src/beampatterns.js
import {abs, range, exp, complex, concat, subtract, multiply, dotMultiply, pi, sum, sqrt, zeros} from "mathjs"

function sinc(ang){
  ang *= pi
  if (Math.abs(ang) < 0.5) {
    // Taylor seris: 1-x^2/3!+x^4/5!-x^6/7!...
    const ang_2 = ang**2;                // x^2
    var curr_term = 1;                   // current term
    var output = 1;                      // the Taylor series result
    for (var ii = 2; ii < 12; ii += 2) {
      curr_term *= -ang_2/(ii*(ii+1))
      output += curr_term
    }
    return output
  } else {
    return Math.sin(ang)/ang
  }
}

function expi(x){
  return(exp(complex(0, x)))
}

class AntennaArray{
  constructor(n_antennas, lambda_ratio){
    this.n_antennas = n_antennas;
    this.lambda_ratio = lambda_ratio;
    this.antenna_index = range(0, n_antennas);
    this.ang_const = 2*lambda_ratio*pi;
  }
  // Angle conversions
  ang_abs2rel(ang){
    return(this.ang_const*Math.sin(ang))
  }
  ang_abs2rel_2(ang){
    var my_ang = ang
    var output = 0
    while (my_ang > pi/2) {
      output += 2*this.ang_const
      my_ang -= pi
    }
    while (my_ang < -pi/2) {
      output -= 2*this.ang_const
      my_ang += pi
    }
    return(output + this.ang_const*Math.sin(my_ang))
  }
  ang_rel2abs_2(ang){
    var my_ang = ang
    var output = 0
    while (my_ang > this.ang_const) {
      output += pi
      my_ang -= 2*this.ang_const
    }
    while (my_ang < -this.ang_const) {
      output -= pi
      my_ang += 2*this.ang_const
    }
    return(output + Math.asin(my_ang/this.ang_const))
  }
  // Steering function
  bp_steer_index(ang){
    return(multiply(subtract(this.antenna_index, (this.n_antennas-1)/2), ang));
  }
  bp_steer(bp, ang){
    const steering_vector = this.bp_steer_index(ang).map(element => expi(element));
    return(dotMultiply(bp, steering_vector));
  }
  // Beam-pattern creation
  bp_sinc_index(width){
    return(multiply(subtract(this.antenna_index, (this.n_antennas-1)/2), width/(2*pi)))
  }
  bp_sinc(width){
    const sinc_index = this.bp_sinc_index(width)
    var bp = sinc_index.map(item => sinc(item))
    const bp_norm = sqrt(sum(bp.map(cc => abs(cc)**2)))
    bp = bp.map(cc => cc/bp_norm)
    return(sinc_index.map(item => sinc(item)))
  }
  // Array response
  set_ang_domain_rel(x){
    this.ang_domain_rel = x;
    const exp_index = x.map(item => multiply(this.antenna_index, item));
    this.response_domain_rel = exp_index.map(item => expi(item));
  }
  array_response_rel(bp){
    return(multiply(this.response_domain_rel, bp));
  }
  set_ang_domain_abs(x){
    this.ang_domain_abs = x;
    const exp_index = x.map(item => multiply(this.antenna_index, this.ang_abs2rel(item)));
    console.log(exp_index);
    this.response_domain_abs = exp_index.map(item => expi(item))
  }
  array_response_abs(bp){
    return(multiply(this.response_domain_abs, bp))
  }
  // Setters
  set_n_antennas(n_antennas){
    this.n_antennas = n_antennas;
    this.antenna_index = range(0, n_antennas);
    if (this.ang_domain_rel) {
      this.set_ang_domain_rel(this.ang_domain_rel)
    }
    if (this.ang_domain_abs) {
      this.set_ang_domain_abs(this.ang_domain_abs)
    }
  }
  set_lambda_ratio(lambda_ratio){
    this.lambda_ratio = lambda_ratio;
    this.ang_const = 2*lambda_ratio*pi;
    if (this.ang_domain_abs) {
      this.set_ang_domain_abs(this.ang_domain_abs)
    }
  }
}

export default AntennaArray

File: python/beampatterns/beampatterns.py
import numpy as np

class AntennaArray:
    def __init__(self, n_antennas=16, lambda_ratio=0.5):
        self.n_antennas = n_antennas
        self.lambda_ratio = lambda_ratio
        self.antenna_index = np.arange(0, n_antennas);
        self.ang_const = 2*lambda_ratio*np.pi;
    # Angle transformations (abs from -pi to np.pi, rel from -2*pi*lambda_ratio to 2*pi*lambda_ratio)
    def ang_abs2rel(self, ang):
        return np.asin(np.array(ang)/(2*self.lambda_ratio*np.pi))
    def ang_abs2rel_2(self, ang):
        my_ang = ang
        output = 0
        while my_ang > np.pi/2:
            output += 2*self.ang_const
            my_ang -= np.pi
        while my_ang < -np.pi/2:
            output -= 2*self.ang_const
            my_ang += np.pi
        return output + self.ang_const*np.sin(my_ang)
    def ang_rel2abs_2(self, ang):
        my_ang = ang
        output = 0
        while my_ang > self.ang_const:
            output += np.pi
            my_ang -= 2*self.ang_const
        while my_ang < -self.ang_const:
            output -= np.pi
            my_ang += 2*self.ang_const
        return output + np.asin(my_ang/self.ang_const)
    # Steering function
    def bp_steer(self, b, ang):
        return [b_ii*np.exp(2j*np.pi*ang) for b_ii, ii in zip(b, range(len(b)))]
    # Beam-pattern creation (relative angle)
    def bp_sinc(self, width):
        half_antennas = (self.n_antennas-1)/2
        return [np.sinc((width/(2*np.pi))*(ii-half_antennas)) for ii in range(self.n_antennas)]
    # Array response
    def set_ang_domain_rel(self, x):
        self.ang_domain_rel = x
        self.response_domain_rel = np.exp(1j*self.antenna_index[:, np.newaxis]*np.array(x)[np.newaxis, :])
    def array_response_rel(self, bp):
        return np.dot(bp, self.response_domain_rel)
    def set_ang_domain_abs(self, x):
        self.ang_domain_rel = x
        self.response_domain_abs = np.exp(1j*self.ang_const*self.antenna_index[:, np.newaxis]*np.sin(x)[np.newaxis, :])
    def array_response_abs(self, bp):
        return np.dot(bp, self.response_domain_abs)
